yellowbox/extras/logstash.py

```diff
@@ -1,30 +1,263 @@
+        """Background thread processing incoming connections and data"""
+        while True:
+            events = self._selector.select(timeout=5)  # For signal handling
+            for key, mask in events:
+                # Handle closing request
+                if key.fileobj is self._rshutdown:
+                    assert self._rshutdown.recv(
+                        len(_CLOSE_SENTINEL)) == _CLOSE_SENTINEL
+                    return
+
+                # Handle new connection
+                if key.fileobj is self._root:
+                    new_socket, _ = self._root.accept()
+                    self._selector.register(new_socket, selectors.EVENT_READ,
+                                            self._create_data_callback(new_socket))
+                    continue
+
+                # Data received, run callback
+                try:
+                    key.data()
+                except Exception:
+                    _logger.exception("Unknown error occurred, closing connection.")
+                    self._selector.unregister(key.fileobj)
+                    # noinspection PyUnresolvedReferences
+                    key.fileobj.close()
+
+    def start(self, *, retry_spec: None = None) -> None:
+        """Start the service.
+
+        Args:
+            retry_spec: Ignored.
+        """
+        self._root.listen(_LISTEN_BACKLOG)
+        self._selector.register(self._root, selectors.EVENT_READ)
+        self._selector.register(self._rshutdown, selectors.EVENT_READ)
+        self._thread.start()
+        super(LogstashService, self).start(retry_spec=retry_spec)
+
+    def stop(self) -> None:
+        """Stop the service
+
+        Raises:
+            RuntimeError: Failed stopping the service. Operation timed out.
+        """
+        if not self._thread.is_alive():
+            return
+        self._wshutdown.send(_CLOSE_SENTINEL)
+        self._thread.join(_STOP_TIMEOUT)  # Should almost never timeout
+        if self._thread.is_alive():
+            raise RuntimeError(f"Failed stopping {self.__class__.__name__}.")
+
+        for key in self._selector.get_map().values():
+            sock = cast(socket.socket, key.fileobj)
+            sock.close()
+
+        self._selector.close()
+
+    def is_alive(self) -> bool:
+        """Check if LogstashService is alive.
+
+        Returns:
+            Boolean.
+        """
+        return self._thread.is_alive()
+
+    def connect(self, network: Any) -> List[str]:
+        """Does nothing. Conforms to YellowService interface."""
+        # Logstash service is not docker related. It cannot actually connect to
+        # the network. However, other containers connected to the network can
+        # connect to the service with docker's usual host
+        return [_docker_host_name]
+
+    def disconnect(self, network: Any):
+        """Does nothing. Conforms to YellowService interface."""
+        pass
+
+    def _filter_records(self, level: int) -> Iterable[Dict[str, Any]]:
```

Comment:
this method should be public, or have a public equivalent

Reply:
I thought about it, wasn't sure if we want one.

Reply:
But you have a method to assert that no logs were received by level? What if I want finer tuning without knowing the logging module extensively?

Reply:
You have one to assert that there are no error logs. Filtering records by level will not help your lack of knowledge as you still need to know the log structure. I'll just make this public as you really want it for some odd reason <3